packages/ui/src/views/w3m-wallet-explorer-view/index.ts

```diff
@@ -89,34 +170,50 @@ export class W3mWalletExplorerView extends LitElement {
 
   // -- render ------------------------------------------------------- //
   protected render() {
-    const { listings } = ExplorerCtrl.state.wallets
+    const { listings } = this.listingResponse
     const classes = {
-      'w3m-loading': this.loading,
+      'w3m-loading': this.isLoadingEndless,
       'w3m-first-fetch': this.firstFetch,
       'w3m-end-reached': this.endReached
     }
 
     return html`
       ${dynamicStyles()}
 
-      <w3m-modal-header title="Wallet Explorer"></w3m-modal-header>
-      <w3m-modal-content class=${classMap(classes)}>
-        <div class="w3m-content">
-          ${listings.map(
-            listing => html`
-              <w3m-wallet-button
-                src=${listing.image_url.lg}
-                name=${listing.name}
-                .onClick=${async () =>
-                  CoreHelpers.isMobile()
-                    ? this.onConnect(listing.mobile, listing.name)
-                    : this.onConnect(listing.desktop, listing.name)}
-              >
-              </w3m-wallet-button>
-            `
-          )}
-        </div>
+      <w3m-modal-header>${this.searchTemplate()}</w3m-modal-header>
 
+      <w3m-modal-content class=${classMap(classes)}>
+        ${this.isLoadingFiltered && !this.isLoadingEndless && !this.firstFetch
+          ? this.loadingTemplate()
+          : html`
+              ${listings.length > 0
+                ? html`
+                    <div class="w3m-content">
+                      ${listings.map(
+                        listing => html`
+                          <w3m-wallet-button
+                            src=${listing.image_url.lg}
+                            name=${listing.name}
+                            .onClick=${async () =>
```

Comment:
Make this a private function with descriptive name. Let's try to keep markup "lean" and logic separate